vars/artifactSetVersion.groovy

```diff
@@ -177,8 +191,82 @@ void call(Map parameters = [:], Closure body = null) {
+                    .use()
+
+                withCredentials([usernamePassword(
+                     credentialsId: config.gitCredentialsId,
+                     passwordVariable: 'PASSWORD',
+                     usernameVariable: 'USERNAME')]) {
+
+                     // Problem: when username/password is encoded and in case the encoded version differs from
+                     // the non-encoded version  (e.g. '@'  gets replaced by '%40' the encoded version
+                     // it is not replaced by stars in the log by surrounding withCredentials.
+                     // In order to avoid having the secrets in the log we take the following actions in case
+                     // the encoded version(s) differs from the non-encoded versions
+                     //
+                     // 1.) we switch off '-x' in the hashbang
+                     // 2.) we tell git push to be silent
+                     // 3.) we send stderr to /dev/null
+                     //
+                     // Disadvantage: In this case we don't see any output for troubleshooting.
+
+                     def USERNAME_ENCODED = URLEncoder.encode(USERNAME, 'UTF-8'),
+                         PASSWORD_ENCODED = URLEncoder.encode(PASSWORD, 'UTF-8')
+
+                     boolean encodedVersionsDiffers = USERNAME_ENCODED != USERNAME || PASSWORD_ENCODED != PASSWORD
+
+                     def prefix = 'https://'
+                     def gitUrlWithCredentials = config.gitHttpsUrl.replaceAll("^${prefix}", "${prefix}${USERNAME_ENCODED}:${PASSWORD_ENCODED}@")
+
+                     def hashbangFlags = '-xe'
+                     def gitPushFlags = []
+                     def streamhandling = ''
+                     gitConfig = []
+
+                     if(config.gitHttpProxy) {
+                         gitConfig.add("-c http.proxy=\"${config.gitHttpProxy}\"")
+                     }
+
+                     if(config.gitDisableSSLVerification) {
+                         echo 'git ssl verification is switched off. This setting is not recommanded in productive environments.'
+                         gitConfig.add('http.sslVerify false')
+                     }
+
+                     if(encodedVersionsDiffers) {
+                         hashbangFlags = '-e'
+                         streamhandling ='&>/dev/null'
+                         gitPushFlags.add('--quiet')
+                         echo 'Performing git push in quiet mode.'
+                     }
+
+                     gitConfig = gitConfig.join(' ')
+                     gitPushFlags = gitPushFlags.join(' ')
+
+                     sh script: """|#!/bin/bash ${hashbangFlags}
+                                   |git ${gitConfig} push ${gitPushFlags} ${gitUrlWithCredentials} ${config.tagPrefix}${newVersion} ${streamhandling}""".stripMargin()
+                 }
+            } else {
+                echo "Git push mode: ${gitPushMode.toString()}. Git push to remote has been skipped."
```

Comment:
Will this work when `gitPushMode` is not set?

Reply:
We will provide a default value on the level of `resources/default_pipeline_environment.yml`. Second answer: we will not fail here in this case but above when the stringish (... in your case `null` is coerced into the enum type.